// brightpearl-client/build/Endpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const CONTACT = "contact-service";
const WAREHOUSE = "warehouse-service";
const ORDER = "order-service";
const ACCOUNTING = "accounting-service";
const PRODUCT = "product-service";
const CONTACT_ID = "contactId";
exports.CONTACT_ID = CONTACT_ID;
const POSTAL_ADDRESS_ID = "postalAddressId";
exports.POSTAL_ADDRESS_ID = POSTAL_ADDRESS_ID;
const WAREHOUSE_ID = "warehouseId";
exports.WAREHOUSE_ID = WAREHOUSE_ID;
const ORDER_ID = "orderId";
exports.ORDER_ID = ORDER_ID;
const ORDER_TYPE = "orderType";
exports.ORDER_TYPE = ORDER_TYPE;
const ORDER_STATUS_ID = "orderStatusId";
exports.ORDER_STATUS_ID = ORDER_STATUS_ID;
const PRODUCT_ID = "productId";
exports.PRODUCT_ID = PRODUCT_ID;
const GOODS_OUT_NOTE_ID = "goodsOutNoteId";
exports.GOODS_OUT_NOTE_ID = GOODS_OUT_NOTE_ID;
const PRICE_LIST_ID = "priceListId";
exports.PRICE_LIST_ID = PRICE_LIST_ID;
const LEAD_SOURCE_ID = "leadSourceId";
exports.LEAD_SOURCE_ID = LEAD_SOURCE_ID;
const PRODUCT_CHANNEL_ID = "productChannelId";
exports.PRODUCT_CHANNEL_ID = PRODUCT_CHANNEL_ID;
const BRAND_ID = "brandId";
exports.BRAND_ID = BRAND_ID;
const CONTACT_TYPE = "contactType";
exports.CONTACT_TYPE = CONTACT_TYPE;
const BRIGHTPEARL_CATEGORY_ID = "brightpearlCategoryId";
exports.BRIGHTPEARL_CATEGORY_ID = BRIGHTPEARL_CATEGORY_ID;
const PRODUCT_TYPE_ID = "productTypeId";
exports.PRODUCT_TYPE_ID = PRODUCT_TYPE_ID;
const PRODUCT_PRICE_ID = "productPriceId";
exports.PRODUCT_PRICE_ID = PRODUCT_PRICE_ID;
const CUSTOM_FIELD_METADATA_ID = "customFieldMetadataId";
exports.CUSTOM_FIELD_METADATA_ID = CUSTOM_FIELD_METADATA_ID;
const PAYMENT_METHOD_ID = "paymentMethodId";
exports.PAYMENT_METHOD_ID = PAYMENT_METHOD_ID;
const ENDPOINTS = Object.freeze({
    contact: {
        search: `${CONTACT}/contact-search`,
        get: `${CONTACT}/contact/{${CONTACT_ID}}`,
        post: `${CONTACT}/contact`,
        patch: `${CONTACT}/contact/{${CONTACT_ID}}`,
        postal_address: {
            put: `${CONTACT}/contact/{${CONTACT_ID}}/postal-address`,
            post: `${CONTACT}/postal-address`,
            get: `${CONTACT}/postal-address/{${POSTAL_ADDRESS_ID}}`
        },
        lead_source: {
            get: `${CONTACT}/lead-source/{${LEAD_SOURCE_ID}}`
        },
        custom_field: {
            get: `${CONTACT}/contact/{${CONTACT_ID}}/custom-field`,
            patch: `${CONTACT}/contact/{${CONTACT_ID}}/custom-field`,
            custom_field_metadata: {
                get: `${CONTACT}/{${CONTACT_TYPE}}/custom-field-meta-data`
            }
        }
    },
    warehouse: {
        base: `${WAREHOUSE}`,
        get: `${WAREHOUSE}/warehouse/{${WAREHOUSE_ID}}`,
        post: `${WAREHOUSE}/warehouse`,
        put: `${WAREHOUSE}/warehouse`,
        goods_out_note: {
            search: `${WAREHOUSE}/goods-note/goods-out-search`,
            post: `${WAREHOUSE}/order/{${ORDER_ID}}/goods-note/goods-out`,
            get: `${WAREHOUSE}/order/{${ORDER_ID}}/goods-note/goods-out/{${GOODS_OUT_NOTE_ID}}`,
            put: `${WAREHOUSE}/goods-note/goods-out/{${GOODS_OUT_NOTE_ID}}`,
            event: {
                post: `${WAREHOUSE}/goods-note/goods-out/{${GOODS_OUT_NOTE_ID}}/event`
            }
        },
        stock_correction: {
            post: `${WAREHOUSE}/warehouse/{${WAREHOUSE_ID}}/stock-correction`
        },
        default_location: {
            get: `${WAREHOUSE}/warehouse/{${WAREHOUSE_ID}}/location/default`
        },
        product_availability: {
            get: `${WAREHOUSE}/product-availability/{${PRODUCT_ID}}`,
        },
        reservation: {
            get: `${WAREHOUSE}/order/{${ORDER_ID}}/reservation`,
            post: `${WAREHOUSE}/order/{${ORDER_ID}}/reservation/warehouse/{${WAREHOUSE_ID}}`
        },
        shipping_method: {
            get: `${WAREHOUSE}/shipping-method`
        }
    },
    order: {
        base: `${ORDER}`,
        get: `${ORDER}/order/{${ORDER_ID}}`,
        search: `${ORDER}/order-search`,
        post: `${ORDER}/order`,
        note: {
            get: `${ORDER}/order/{${ORDER_ID}}/note`,
            post: `${ORDER}/order/{${ORDER_ID}}/note`,
            search: `${ORDER}/order-note-search`
        },
        row: {
            post: `${ORDER}/order/{${ORDER_ID}}/row`
        },
        sales: {
            post: `${ORDER}/sales-order`,
            get: `${ORDER}/sales-order/{${ORDER_ID}}`,
            close: {
                post: `${ORDER}/sales-order/{${ORDER_ID}}/close`
            }
        },
        status: {
            get: `${ORDER}/order-status/{${ORDER_STATUS_ID}}`,
            put: `${ORDER}/order/{${ORDER_ID}}/status`
        },
        custom_field: {
            patch: `${ORDER}/order/{${ORDER_ID}}/custom-field`,
            custom_field_metadata: {
                get: `${ORDER}/{${ORDER_TYPE}}/custom-field-meta-data`
            },
            get: `${ORDER}/order/{${ORDER_ID}}/custom-field`
        },
        order_type: {
            get: `${ORDER}/order-type`
        }
    },
    sales_credit: {
        get: `${ORDER}/sales-credit/{${ORDER_ID}}`
    },
    accounting: {
        customer_payment: {
            post: `${ACCOUNTING}/customer-payment`,
            search: `${ACCOUNTING}/customer-payment-search`
        },
        invoice: {
            sales: {
                post: `${ACCOUNTING}/invoice/sales-invoice`
            }
        },
        currency: {
            search: `${ACCOUNTING}/currency-search`
        },
        payment: {
            search: `${ACCOUNTING}/payment-method-search`,
            get: `${ACCOUNTING}/payment-method/{${PAYMENT_METHOD_ID}}`
        }
    },
    product: {
        search: `${PRODUCT}/product-search`,
        get: `${PRODUCT}/product/{${PRODUCT_ID}}`,
        post: `${PRODUCT}/product`,
        put: `${PRODUCT}/product/{${PRODUCT_ID}}`,
        supplier: {
            get: `${PRODUCT}/product/{${PRODUCT_ID}}/supplier`,
            post: `${PRODUCT}/product/{${PRODUCT_ID}}/supplier`,
            delete: `${PRODUCT}/product/{${PRODUCT_ID}}/supplier`
        },
        price_list: {
            get: `${PRODUCT}/price-list/{${PRICE_LIST_ID}}`
        },
        channel: {
            get: `${PRODUCT}/channel/{${PRODUCT_CHANNEL_ID}}`
        },
        brand: {
            search: `${PRODUCT}/brand-search`,
            get: `${PRODUCT}/brand/{${BRAND_ID}}`,
            post: `${PRODUCT}/brand`
        },
        brightpearl_category: {
            search: `${PRODUCT}/brightpearl-category-search`,
            get: `${PRODUCT}/brightpearl-category/{${BRIGHTPEARL_CATEGORY_ID}}`,
            post: `${PRODUCT}/brightpearl-category`
        },
        product_type: {
            search: `${PRODUCT}/product-type-search`,
            get: `${PRODUCT}/product-type/{${PRODUCT_TYPE_ID}}`,
            post: `${PRODUCT}/product-type`
        },
        product_price: {
            get_all: `${PRODUCT}/product-price/{${PRODUCT_ID}}`,
            get: `${PRODUCT}/product-price/{${PRODUCT_ID}}/price-list/{${PRODUCT_PRICE_ID}}`,
            put: `${PRODUCT}/product-price/{${PRODUCT_ID}}/price-list`
        },
        custom_field: {
            patch: `${PRODUCT}/product/{${PRODUCT_ID}}/custom-field`,
            custom_field_metadata: {
                get: `${PRODUCT}/product/custom-field-meta-data/{${CUSTOM_FIELD_METADATA_ID}}`
            }
        },
        product_custom_field: {
            get: `${PRODUCT}/product-custom-field/{${PRODUCT_ID}}`
        }
    }
});
exports.ENDPOINTS = ENDPOINTS;
