// brightpearl-client/build/IBrightpearlClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var PostalAddressTypes;
(function (PostalAddressTypes) {
    PostalAddressTypes["DEFAULT"] = "DEF";
    PostalAddressTypes["BILLING"] = "BIL";
    PostalAddressTypes["DELIVERY"] = "DEL";
})(PostalAddressTypes || (PostalAddressTypes = {}));
exports.PostalAddressTypes = PostalAddressTypes;
var PaymentType;
(function (PaymentType) {
    PaymentType["AUTH"] = "AUTH";
    PaymentType["CAPTURE"] = "CAPTURE";
    PaymentType["PAYMENT"] = "PAYMENT";
    PaymentType["RECEIPT"] = "RECEIPT";
    PaymentType["VOID"] = "VOID";
})(PaymentType || (PaymentType = {}));
exports.PaymentType = PaymentType;
var GoodsOutNoteEventCodes;
(function (GoodsOutNoteEventCodes) {
    GoodsOutNoteEventCodes["PLA"] = "PLA";
    GoodsOutNoteEventCodes["PRI"] = "PRI";
    GoodsOutNoteEventCodes["UPR"] = "UPR";
    GoodsOutNoteEventCodes["PIC"] = "PIC";
    GoodsOutNoteEventCodes["UPI"] = "UPI";
    GoodsOutNoteEventCodes["PAC"] = "PAC";
    GoodsOutNoteEventCodes["UPA"] = "UPA";
    GoodsOutNoteEventCodes["SHW"] = "SHW";
})(GoodsOutNoteEventCodes || (GoodsOutNoteEventCodes = {}));
exports.GoodsOutNoteEventCodes = GoodsOutNoteEventCodes;
var OrderType;
(function (OrderType) {
    OrderType["PURCHASE_ORDER"] = "purchase";
    OrderType["SALE_ORDER"] = "sale";
})(OrderType || (OrderType = {}));
exports.OrderType = OrderType;
var OrderTypeCode;
(function (OrderTypeCode) {
    OrderTypeCode["SALES_ORDER"] = "SO";
    OrderTypeCode["PURCHASE_ORDER"] = "PO";
    OrderTypeCode["SALES_CREDIT"] = "SC";
    OrderTypeCode["PURCHASE_CREDIT"] = "PC";
})(OrderTypeCode || (OrderTypeCode = {}));
exports.OrderTypeCode = OrderTypeCode;
var CustomFieldOperation;
(function (CustomFieldOperation) {
    CustomFieldOperation["ADD"] = "add";
    CustomFieldOperation["REPLACE"] = "replace";
    CustomFieldOperation["REMOVE"] = "remove";
    CustomFieldOperation["MOVE"] = "move";
    CustomFieldOperation["TEST"] = "test";
})(CustomFieldOperation || (CustomFieldOperation = {}));
exports.CustomFieldOperation = CustomFieldOperation;
var AllOrders;
(function (AllOrders) {
    AllOrders["ORDERID"] = "*";
})(AllOrders || (AllOrders = {}));
exports.AllOrders = AllOrders;
var OrderUrlParams;
(function (OrderUrlParams) {
    OrderUrlParams["CUSTOM_FIELDS"] = "customFields";
    OrderUrlParams["NULL_CUSTOM_FIELDS"] = "nullCustomFields";
})(OrderUrlParams || (OrderUrlParams = {}));
exports.OrderUrlParams = OrderUrlParams;
var ContactUrlParams;
(function (ContactUrlParams) {
    ContactUrlParams["CUSTOM_FIELDS"] = "customFields";
    ContactUrlParams["NULL_CUSTOM_FIELDS"] = "nullCustomFields";
    ContactUrlParams["POSTAL_ADDRESSES"] = "postalAddresses";
})(ContactUrlParams || (ContactUrlParams = {}));
exports.ContactUrlParams = ContactUrlParams;
var ProductAvailabilityUrlParams;
(function (ProductAvailabilityUrlParams) {
    ProductAvailabilityUrlParams["BREAK_DOWN_BY_LOCATION"] = "breakDownByLocation";
    ProductAvailabilityUrlParams["ALLOCATED_ORDERS"] = "allocatedOrders";
    ProductAvailabilityUrlParams["ALLOCATED_TRANSFERS"] = "allocatedTransfers";
})(ProductAvailabilityUrlParams || (ProductAvailabilityUrlParams = {}));
exports.ProductAvailabilityUrlParams = ProductAvailabilityUrlParams;
var WarehouseType;
(function (WarehouseType) {
    WarehouseType["CONCESSION"] = "concession";
    WarehouseType["CONSIGNMENT"] = "consignment";
    WarehouseType["MISCELLANEOUS"] = "miscellaneous";
    WarehouseType["ONLINE"] = "online";
    WarehouseType["STORE"] = "store";
    WarehouseType["SUPPLIER"] = "supplier";
    WarehouseType["THIRD_PARTY_LOGISTICS"] = "3pl";
    WarehouseType["VIRTUAL"] = "virtual";
    WarehouseType["WAREHOUSE"] = "warehouse";
})(WarehouseType || (WarehouseType = {}));
exports.WarehouseType = WarehouseType;
